Restructured and organized; address issue 1252 (#3219)
mlflow/mlflow: CONTRIBUTING.rst

After you have agreed upon an implementation strategy for your feature or patch with an MLflow
committer, the next step is to introduce your changes (see `developing changes

> <https://github.com/mlflow/mlflow/blob/master/CONTRIBUTING.rst#developing-and-testing-changes-for-mlflow>`_)

Reviewer: If we stick with the original wording per https://github.com/mlflow/mlflow/pull/3219/files#r472528940, we should undo this change

Reviewer: Given that we're ok with the wording "Developing and testing MLflow" as the header, can we fix this link? I think it should be https://github.com/mlflow/mlflow/blob/master/CONTRIBUTING.rst#developing-and-testing-mlflow, to match https://github.com/mlflow/mlflow/blob/ebacfdaf0095bf816a19f24a64319d6570d81f26/CONTRIBUTING.rst#developing-and-testing-mlflow

For more information about Plugins, see https://mlflow.org/docs/latest/plugins.html.

> Developing and testing MLflow

Reviewer: IMO the original wording here is actually a bit clearer, is there a reason to change it? With "Developing and testing MLflow" it's a bit less clear that we're testing changes to MLflow, as opposed to using MLflow to develop another project

Author: One argument would be while the proposition "to" or "for" makes it specific to MLflow, we are already talking about contributions to MLFlow. The document is about contributing to MLflow. So it may be just redundant. However, the CONTRIBUTIONS.RST was for MLflow and say, Spark MLlib, then having that specificity would make it more clear.

Reviewer: Gotcha, yeah seems fine given that we explicitly state that this page is for "contributing to MLflow" at the top

> Developing and testing changes for MLflow

Author: I suppose you can drop the proposition.

#############################
The majority of the MLflow codebase is developed in Python. This includes the CLI, Tracking Server,
Artifact Repositories (e.g., S3 or Azure Blob Storage backends), and of course the Python fluent,
tracking, and model APIs.

> General prerequisites and dependencies

Reviewer: Any reason to reword this? The original wording seemed a bit more concise

Author: I had taught about general (applies to all) and specific to individual components. For example, git applies to all; installing JavaScript dependencies is more specific to that component.

Reviewer: Ah got it, sorry makes sense in the context of the surrounding content - maybe "Common prerequisites and dependencies"

Author: Yes, "Common prerequisites and dependencies" makes sense

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
First, ensure that your name and email are
`configured in git <https://git-scm.com/book/en/v2/Getting-Started-First-Time-Git-Setup>`_ so that

> you can `sign your work`_ when committing code changes and opening pull requests:

Reviewer 2: Do you think it would be good to say why we ask them to ask their work?

Author: We mention that in the[ Sign your work](https://github.com/mlflow/mlflow/blob/master/CONTRIBUTING.rst#sign-your-work) section: " You need to sign that you wrote the patch or otherwise have the right to pass it on as an open-source patch."

Ensure `Docker <https://www.docker.com/>`_ is installed.

> Finally, we use ``pytest`` to run all Python contributed code. Install ``pytest``:

Author: Yes, more precise.

.. code-block:: bash

    pip install pytest

JavaScript and UI
~~~~~~~~~~~~~~~~~

> ``npm`` is required to run the Javascript dev server and the tracking UI.

Reviewer 2: For completeness, can we add an introduction like : 
The MLflow UI is written in JavaScript. `npm` ...

Author: That sounds reasonable.

If modifying dependencies in ``mlflow/server/js/package.json``, run ``npm update`` within
``mlflow/server/js`` to install the updated dependencies.

> Launching the Development UI

Reviewer 2: Do you think we need to have separate headings for Launching the Dev UI and Running the JS Dev Server. Can they be merged?

Author: The original document was written by @smurching, and when we restructured it we placed it as such. I'll leave that up to @smurching if it makes sense to merge them.

install `Java <https://www.java.com/>`_ and `Apache Maven <https://maven.apache.org/download.cgi>`_.

Certain MLflow modules are implemented in Java, under the ``mlflow/java/`` directory.
These are the Java Tracking API client (``mlflow/java/client``) and the Model Scoring Server
for Java-based models like MLeap (``mlflow/java/scoring``).

Other Java functionality (like artifact storage) depends on the Python package, so first install
the Python package in a conda environment as described above.
`Install <https://www.oracle.com/technetwork/java/javase/downloads/index.html>`_
the Java 8 JDK (or above), and `download <https://maven.apache.org/download.cgi>`_
and `install <https://maven.apache.org/install.html>`_ Maven. You can then build and run tests via:

.. code-block:: bash

  cd mlflow/java
  mvn compile test

If opening a PR that makes API changes, please regenerate API documentation as described in
`Writing Docs`_ and commit the updated docs to your PR branch.

Python
~~~~~~

Majority of contributions to MLflow are in Python, and we use ``pytest`` to run all Python contributed code. Install ``pytest``:

.. code-block:: bash

>     pip install pytest

Reviewer 3: Should we remove this block because pytest should be installed in the "General prerequisites and dependencies" step?

Author: We could since you will need it for Python, in general, so might as well. Will do it.

Author: done!

Certain MLflow modules are implemented in Java, under the ``mlflow/java/`` directory.
These are the Java Tracking API client (``mlflow/java/client``) and the Model Scoring Server
for Java-based models like MLeap (``mlflow/java/scoring``).

Other Java functionality (like artifact storage) depends on the Python package, so first install
the Python package in a conda environment as described above.
`Install <https://www.oracle.com/technetwork/java/javase/downloads/index.html>`_
the Java 8 JDK (or above), and `download <https://maven.apache.org/download.cgi>`_
and `install <https://maven.apache.org/install.html>`_ Maven. You can then build and run tests via:

.. code-block:: bash

  cd mlflow/java
  mvn compile test

If opening a PR that makes API changes, please regenerate API documentation as described in
`Writing Docs`_ and commit the updated docs to your PR branch.

Python
~~~~~~
If you are contributing in Python, make sure that you have satisfied all the `General prerequisites and dependencies`_,
including installing ``pytest``, as you will need it for the sections described below.

> Writing Python Tests

Reviewer 2: Can the section related to writing and running python test come after the Python Client and Python Server? Makes a little more sense to me but open to suggestions!

Author: As they apply to both client and server, we decided to put them before.

++++++++++++++++++++
If your PR includes code that isn't currently covered by our tests (e.g. adding a new flavor, adding
autolog support to a flavor, etc.), you should write tests that cover your new code. MLflow currently
uses ``pytest==3.2.1`` for testing. Your tests should be added to the relevant file under ``tests``, or
if there is no appropriate file, in a new file prefixed with ``test_`` so that ``pytest`` includes that
file for testing.

If your tests require usage of a tracking URI, the
`pytest fixture <https://docs.pytest.org/en/3.2.1/fixture.html>`_
`tracking_uri_mock <https://github.com/mlflow/mlflow/blob/master/tests/conftest.py#L74>`_ is automatically set up
for every tests. It sets up a mock tracking URI that will set itself up before your test runs and tear itself down after.
If you want to deactivate the mock for your test, mark the test with `@pytest.mark.notrackingurimock` operator.

Running Python Tests
++++++++++++++++++++

Verify that the unit tests & linter pass before submitting a pull request by running:

.. code-block:: bash

    ./lint.sh
    ./dev/run-small-python-tests.sh

>     # Optionally, run large tests as well. Travis will run large tests on your pull request once

Reviewer 2: I dont think travis runs the tests anymore. Github actions for the win!
```suggestion
    # Optionally, run large tests as well. Github actions will run large tests on your pull request once
```

Author: There places where we will have to change all occurrences of how Travis is replaced with Github's actions.

Please also follow the recommendations from the
`Advanced R - Style Guide <http://adv-r.had.co.nz/Style.html>`_ regarding naming and styling.

Java
~~~~

If contributing to MLflow's Java APIs or modifying Java documentation,
install `Java <https://www.java.com/>`_ and `Apache Maven <https://maven.apache.org/download.cgi>`_.

Certain MLflow modules are implemented in Java, under the ``mlflow/java/`` directory.
These are the Java Tracking API client (``mlflow/java/client``) and the Model Scoring Server
for Java-based models like MLeap (``mlflow/java/scoring``).

> Other Java functionality (like artifact storage) depends on the Python package, so first install

Reviewer: Can we replace this sentence ("Other Java functionality (like artifact storage) depends on the Python package, so first install the Python package in a conda environment as described above.") with a link to the "Common prerequisites and dependencies" section, as you've done for Python [here](https://github.com/mlflow/mlflow/pull/3219/files#diff-ea1f65a8eaac0cec7ae151dfdbbbc587R292)? And could we do the same for R?

Author: sounds good; will do it.

the Python package in a conda environment as described above.
`Install <https://www.oracle.com/technetwork/java/javase/downloads/index.html>`_
the Java 8 JDK (or above), and `download <https://maven.apache.org/download.cgi>`_
and `install <https://maven.apache.org/install.html>`_ Maven. You can then build and run tests via:

.. code-block:: bash

  cd mlflow/java
  mvn compile test

If opening a PR that makes API changes, please regenerate API documentation as described in
`Writing Docs`_ and commit the updated docs to your PR branch.

Python
~~~~~~
If you are contributing in Python, make sure that you have satisfied all the `General prerequisites and dependencies`_,
including installing ``pytest``, as you will need it for the sections described below.

Writing Python Tests
++++++++++++++++++++
If your PR includes code that isn't currently covered by our tests (e.g. adding a new flavor, adding

> autolog support to a flavor, etc.), you should write tests that cover your new code. MLflow currently

Reviewer: Can we delete the section "MLflow currently uses ``pytest==3.2.1`` for testing?" Seems unlikely to be immediately useful to readers of this guide & hard to keep up-to-date

Author: Okay, done!

uses ``pytest==3.2.1`` for testing. Your tests should be added to the relevant file under ``tests``, or
if there is no appropriate file, in a new file prefixed with ``test_`` so that ``pytest`` includes that
file for testing.

If your tests require usage of a tracking URI, the
`pytest fixture <https://docs.pytest.org/en/3.2.1/fixture.html>`_
`tracking_uri_mock <https://github.com/mlflow/mlflow/blob/master/tests/conftest.py#L74>`_ is automatically set up
for every tests. It sets up a mock tracking URI that will set itself up before your test runs and tear itself down after.

> If you want to deactivate the mock for your test, mark the test with `@pytest.mark.notrackingurimock` operator.

Reviewer: ```suggestion
By default, runs are logged under a local temporary directory that's unique to each test and torn down immediately after test execution. To disable this behavior, decorate your test function with ``@pytest.mark.notrackingurimock``
```

Author: done!

.. code-block:: bash

    ./lint.sh
    ./dev/run-small-python-tests.sh
    # Optionally, run large tests as well. Github actions will run large tests on your pull request once
    # small tests pass. Note: models and model deployment tests are considered "large" tests. If
    # making changes to these components, we recommend running the relevant tests (e.g. tests under
    # tests/keras for changes to Keras model support) locally before submitting a pull request.
    ./dev/run-large-python-tests.sh

Python tests are split into "small" & "large" categories, with new tests falling into the "small"
category by default. Tests that take 10 or more seconds to run should be marked as large tests

> via the @pytest.mark.large annotation. Dependencies for small and large tests can be added to

Reviewer: ```suggestion
via the ``@pytest.mark.large`` annotation. Dependencies for small and large tests can be added to
```

Sorry to add new suggestions, this just seemed like a good one to fix

Author: No worries, good to fix it. Done!

If opening a PR that changes or adds new APIs, please update or add Python documentation as
described in `Writing Docs`_ and commit the docs to your PR branch.

Python Client
+++++++++++++

For the client, if you are adding new model flavors, follow the instructions below.

Python Model Flavors
--------------------

If you are adding new framework flavor support, you'll need to modify ``pytest`` and Travis configurations so tests for your code can run properly. Generally, the files you'll have to edit are:

> 1. ``.travis.yml``: exclude your tests in the Windows bash script

Reviewer 2: Remove reference to travis here and the line above.

Author: Will delete this line if we don't use it. But there's also reference to `NOT_CRAN=true LINTR_COMMENT_BOT=false Rscript ../.travis.R`. What to do about that one?
Any suggestions?

Author: done!

> 4. ``travis/large-requirements.txt``: add your framework and version to the list of requirements

Reviewer 3: We shoul rename `traivs` folder because we no longer use it. I'll make a PR.

Author: Okay, I'll wait for your PR to make that change since I can't rename it until your PR is merged.

Reviewer 3: Addressed by #3253

      Generating ~/mlflow/mlflow/store/db_migrations/versions/b446d3984cfa_add_new_field_to_db.py

These commands generate a new migration script (e.g., at ``~/mlflow/mlflow/alembic/versions/12341123_add_new_field_to_db.py``)
that you should then edit to add migration logic.

> Writing MLflow Examples

Reviewer 2: I dont think we need to nest this heading under Python. Let me know what you think!

Author: Most of these examples are in Python, and @smurching saw it as fitting to put it under Python, at least for now.

Reviewer: Oops, sorry - agree that it makes sense for examples to go into their own top-level section, i.e. "Developing and testing MLflow -> Writing MLflow examples", sorry for the confusion there.

Reviewer: Looks like if a reader follows the "general prereqs" step, they'll actually install pytest & so be able to run the examples tests / develop examples

Author: Excellent! Will move it under "Developing and testing MLflow"

+++++++++++++++++++++++
The ``mlflow/examples`` directory has a collection of quickstart tutorials and various simple examples that depict MLflow tracking,
project, model flavors, model registry, and serving use cases. Written in `PEP8 Python coding style <https://www.python.org/dev/peps/pep-0008/>`_, these
examples provide developers sample code, as a quick way to learn MLflow Python APIs.

To ease your contributions’ review process, keep examples short yet demonstrative, add documentation, peruse through
some ``mlflow/examples`` as templates for your contribution, and follow the recommended steps below.

If you are contributing by adding new model flavor, follow these steps:

1. Follow instructions in `Python Model Flavors`_
2. Create a corresponding directory in ``mlflow/examples/new-model-flavor``
3. Implement your Python training ``new-model-flavor`` code in this directory

> 4. Convert this directory’s content into an MLflow Project executable

Reviewer 2: Can we add a link for a new user to understand what this means? They may not know what is MLflow project, etc

Author: good point! will do

      Generating ~/mlflow/mlflow/store/db_migrations/versions/b446d3984cfa_add_new_field_to_db.py

These commands generate a new migration script (e.g., at ``~/mlflow/mlflow/alembic/versions/12341123_add_new_field_to_db.py``)
that you should then edit to add migration logic.

Writing MLflow Examples
+++++++++++++++++++++++
The ``mlflow/examples`` directory has a collection of quickstart tutorials and various simple examples that depict MLflow tracking,

> project, model flavors, model registry, and serving use cases. Written in `PEP8 Python coding style <https://www.python.org/dev/peps/pep-0008/>`_, these

Reviewer: Maybe we can remove this reference to PEP8 coding style, as @harupy's linting work now ensures that we lint our examples

Author: cool, will do!

examples provide developers sample code, as a quick way to learn MLflow Python APIs.

> To ease your contributions’ review process, keep examples short yet demonstrative, add documentation, peruse through

Reviewer: ```suggestion
To facilitate review, strive for brief examples that reflect real user workflows, document how to run your example, and follow the recommended steps below.
```

Author: I like that! done!

some ``mlflow/examples`` as templates for your contribution, and follow the recommended steps below.

> If you are contributing by adding new model flavor, follow these steps:

Reviewer: ```suggestion
If you are contributing a new model flavor, follow these steps:
```

Author: done!

1. Follow instructions in `Python Model Flavors`_
2. Create a corresponding directory in ``mlflow/examples/new-model-flavor``
3. Implement your Python training ``new-model-flavor`` code in this directory
4. Convert this directory’s content into an `MLflow Project <https://mlflow.org/docs/latest/projects.html>_` executable
5. Add ``README.md``, ``MLproject``, and ``conda.yaml`` files and your code
6. Read instructions in the ``mlflow/test/examples/README.md`` and add a ``pytest`` entry in the ``test/examples/test_examples.py``
7. Add a short description in the ``mlflow/examples/README.md`` file

If you are contributing to the quickstart directory, follow these steps:

1. Create a descriptive Python program file with --help options
2. Add a short description in the ``mlflow/examples/README.md`` file

> 3. Read instructions in the ``mlflow/test/examples/README.md``, and add a ``pytest`` entry in the ``test/examples/test_examples.py``

Reviewer: We don't actually want to encourage people to add new code to the quickstart (editing the existing quickstart is fine, but we should keep it brief) - maybe we should remove this section? Or just have a sentence stating that we welcome changes to the quickstart that make it clearer or simpler etc.

> If you are contributing to demonstrate a general MLflow functionality that is neither a quickstart nor a new model flavor, follow these steps:

Reviewer: ```suggestion
If you'd like to provide an example of functionality that doesn't fit into the above categories, follow these steps:
```

Author: done!
